### PR TITLE
feature: exclude from recents

### PR DESCRIPTION
**Pull request by sgpublic**
re-push #281

### CODE REVIEW

**Comment by jmattheis** [comment]
Hey, I'm currently a little busy, so it may take some time until I'll have a look at this PR.

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -34,6 +34,7 @@
             android:name=".messages.MessagesActivity"
             android:exported="false"
             android:launchMode="singleTask"
+            android:taskAffinity=".messages.MessagesActivity"
```
Comment (jmattheis) [inline app/src/main/AndroidManifest.xml:37]: Why is thin needed?

File: app/src/main/kotlin/com/github/gotify/init/InitializationActivity.kt [hunk 2]
```diff
@@ -115,8 +115,8 @@ internal class InitializationActivity : AppCompatActivity() {
         settings.setUser(user.name, user.isAdmin)
         requestVersion {
             splashScreenActive = false
-            startActivity(Intent(this, MessagesActivity::class.java))
-            finish()
+            startActivity(MessagesActivity.createIntent(this))
+            finishAndRemoveTask()
```
Comment (jmattheis) [inline app/src/main/kotlin/com/github/gotify/init/InitializationActivity.kt:119]: Does using only `finish` cause problems with the new feature?

File: app/src/main/kotlin/com/github/gotify/settings/SettingsActivity.kt [hunk 3]
```diff
@@ -121,13 +128,15 @@ internal class SettingsActivity : AppCompatActivity(), OnSharedPreferenceChangeL
...
+                task.finishAndRemoveTask()
+            }
             val packageManager = requireContext().packageManager
             val packageName = requireContext().packageName
             val intent = packageManager.getLaunchIntentForPackage(packageName)
             val componentName = intent!!.component
             val mainIntent = Intent.makeRestartActivityTask(componentName)
             startActivity(mainIntent)
-            Runtime.getRuntime().exit(0)
         }
```
Comment (jmattheis) [inline app/src/main/kotlin/com/github/gotify/settings/SettingsActivity.kt:140]: What does this change do?, why do we have to remove the appTasks here?

File: app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (sgpublic) [reply app/src/main/AndroidManifest.xml:37]: According to the instructions in [Android Developer: excludeFromRecent](https://developer.android.com/guide/topics/manifest/activity-element#exclude), if excludeFromRecent wants to take effect, the new activity must be created in a new task, so taskAffinity is added.

File: app/src/main/kotlin/com/github/gotify/init/InitializationActivity.kt (hunk 2 above)
Comment (sgpublic) [reply app/src/main/kotlin/com/github/gotify/init/InitializationActivity.kt:119]: According to my test, if you just kill the process here, you cannot end all tasks, which will cause excludeFromRecent to fail to take effect correctly, so all tasks must be ended here.

File: app/src/main/kotlin/com/github/gotify/settings/SettingsActivity.kt (hunk 3 above)
Comment (sgpublic) [reply app/src/main/kotlin/com/github/gotify/settings/SettingsActivity.kt:140]: To be honest, I think it is the simplest and most efficient way to directly modify the visibility with ActivityManager at the beginning.

File: app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (jmattheis) [reply app/src/main/AndroidManifest.xml:37]: So we could add the `FLAG_ACTIVITY_NEW_TASK` to the MessageActivity intent, to get the same behavior?

Comment (sgpublic) [reply app/src/main/AndroidManifest.xml:37]: But according to my test, FLAG_ACTIVITY_NEW_TASK doesn't really work here

File: app/src/main/kotlin/com/github/gotify/settings/SettingsActivity.kt (hunk 3 above)
Comment (jmattheis) [reply app/src/main/kotlin/com/github/gotify/settings/SettingsActivity.kt:140]: You mean like it was done in your first PR? I think I agree, now this feature could affect other parts of Gotify and your first change only affected the app, if the feature flag was enabled. Do you still have the changes from your first PR? The old PR https://github.com/gotify/android/pull/281 seems broken and doesn't have changes anymore.

Comment (sgpublic) [reply app/src/main/kotlin/com/github/gotify/settings/SettingsActivity.kt]: Yes, I have forced push the change in #281 to this PR.

Comment (jmattheis) [reply app/src/main/kotlin/com/github/gotify/settings/SettingsActivity.kt]: Looks good, I changed the signature of the method, to not be an extension method. And the implementation can use getOrNull, to remove the need for the takeIf.

**Review by jmattheis — APPROVED** [review]

File: app/src/main/kotlin/com/github/gotify/settings/SettingsActivity.kt (hunk 3 above)
Comment (sgpublic) [reply app/src/main/kotlin/com/github/gotify/settings/SettingsActivity.kt]: Thanks for your work! ❤️

**Pull request merged**